src/ragger/navigator/navigator.py

```diff
@@ -394,51 +406,48 @@ def navigate_until_text_and_compare(self,
 
         start = time()
 
-        # Retrieve the screen context.
-        # On Speculos backend it immediately returns without any wait.
-        ctx = self._backend.wait_for_screen_change(0.1)
-
-        # Make sure to enter the navigation loop after
-        # there is a screen change (or wait 2 seconds).
-        # Useful when the navigate_until_text is used
-        # in transaction flows and the screen changes
-        # from the idle menu to a review start screen.
-        try:
-            ctx = self._backend.wait_for_screen_change(2.0, ctx)
-        except TimeoutError:
-            # Maybe the start screen of the transaction flow is
-            # already displayed so we'll wait 2.0 sec until
-            # the exception is thrown then we pass and go to the
-            # navigation loop.
-            pass
+        if screen_change_before_first_instruction:
+            self._backend.wait_for_screen_change(timeout)
 
         # Navigate until the text specified in argument is found.
         while True:
-            if idx is not None:
-                assert isinstance(snaps_tmp_path, Path)
-                assert isinstance(snaps_golden_path, Path)
+            if snaps_tmp_path and snaps_golden_path:
                 self._compare_snap(snaps_tmp_path, snaps_golden_path, idx)
-                idx += 1
 
             if not self._backend.compare_screen_with_text(text):
                 # Go to the next screen.
                 self.navigate([navigate_instruction])
+                idx += 1
             else:
-                # Validation action when the text is found.
-                self.navigate([validation_instruction])
+                # Validation screen text found, exit the loop
                 break
 
             # Global navigation loop timeout in case the text is never found.
-            if (time() - start > timeout):
+            remaining = timeout - (time() - start)
+            if (remaining < 0):
                 raise TimeoutError(f"Timeout waiting for text {text}")
 
-            ctx = self._backend.wait_for_screen_change(2.0, ctx)
+            self._backend.wait_for_screen_change(remaining)
+
+        # Perform navigation validation instructions in an "navigate_and_compare" way.
+        if validation_instructions:
```

Comment:
Do you need this check ? `validation_instructions` is a mandatory parameter.

Reply:
It could be an empty list. In the case where it's called from `navigate_until_text` but there is nothing that should be done once the screen is found.

Reply:
True.